### ISSUE TITLE
Refresh EKS tokens.

### ISSUE THREAD

**Issue by brycechesternewman** [hagay3/skuber#150]
The skuber client appears to have no way to refresh EKS tokens - unless I am missing how?
If I am not missing how to refresh tokens we need a way to do this.

The following is from AWS....

"Hello,

We have identified applications running in one or more of your Amazon EKS clusters that are not refreshing service account tokens. Applications making requests to Kubernetes API server with expired tokens will fail. You can resolve the issue by updating your application and its dependencies to use newer versions of Kubernetes client SDK that automatically refreshes the tokens. 

What is the problem?

Kubernetes version 1.21 graduated BoundServiceAccountTokenVolume feature [1] to beta and enabled it by default. This feature improves security of service account tokens by requiring a one hour expiry time, over the previous default of no expiration. This means that applications that do not refetch service account tokens periodically will receive an HTTP 401 unauthorized error response on requests to Kubernetes API server with expired tokens. You can learn more about the BoundServiceAccountToken feature in EKS Kubernetes 1.21 release notes [2].

To enable a smooth migration of applications to the newer time-bound service account tokens, EKS v1.21+ extends the lifetime of service account tokens to 90 days. Applications on EKS v1.21+ clusters that make API server requests with tokens that are older than 90 days will receive an HTTP 401 unauthorized  error response.

How can you resolve the issue?

To make the transition to time bound service account tokens easier, Kubernetes has updated the below official versions of client SDKs to automatically refetch tokens before the one hour expiration:

* Go v0.15.7 and later
* Python v12.0.0 and later
* Java v9.0.0 and later
* Javascript v0.10.3 and later
* Ruby master branch
* Haskell v0.3.0.0

We recommend that you update your application and its dependencies to use one of the above client SDK versions if you are on an older version. 

While not an exhaustive list, the below AWS components have been updated to use the newer Kubernetes client SDKs that automatically refetches the token :

* Amazon VPC CNI: v1.8.0 and later
* CoreDNS: v1.8.4 and later
* AWS Load Balancer Controller: v2.0.0 and later
* kube-proxy: v1.21.2-eksbuild.2 and later 

You can find instructions for upgrading Amazon VPC CNI, CoreDNS, AWS Load Balancer Controller and kube-proxy in EKS documentation at [3], [4], [5] and [6] respectively. Please check EKS documentation [2] for list of components and the versions that have addressed this issue. 

How can I identify service accounts using stale tokens?

As of April 20th 2022, we have identified the below service accounts attached to pods in one or more of your EKS clusters using stale (older than 1 hour) tokens. Service accounts are listed in the format : <eks-cluster-arn>|<namespace:serviceaccount>

...<ommited>"

**Comment by hagay3** [comment]
@brycechesternewman 
Thank you for creating this issue.
The ability to refresh EKS token is not yet implemented in the library.
But for GCP cluster the refresh operation is already implemented.
https://github.com/hagay3/skuber/blob/master/client/src/main/scala/skuber/api/client/package.scala#L102

I will consider working on this PR and implementing it in skuber.

You can still create a service account with EKS cluster, which does not have an expiration date.
https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html

After creating the service account, you will have a service account associated with secret.
When you have the secret in k8s you can fetch its name with the following command:
```
kubectl get sa
kubectl get sa my-admin-service-account -o yaml
```
you will find the secret name there.
then fetch the token from the secret:
```
kubectl get secret secret-42h5m -o yaml
```

Copy the ```ca.crt``` value.
Copy the ```token``` value.

```
val cluster = Cluster(
      apiVersion = "v1",
      server = "https://CLUSTER_URL",
      insecureSkipTLSVerify = false,
      certificateAuthority = Some(Right(Base64.getDecoder.decode("THE_CA.CRT_VALUE_COPIED")))
    )
    val context = Context(cluster = cluster, authInfo = TokenAuth(new String(Base64.getDecoder.decode("THE_TOKEN_COPIED"))))
    val k8sConfig = Configuration().withCluster("cluster", cluster).withContext("cluster", context).useContext(context)
    val k8sClient: KubernetesClient = k8sInit(k8sConfig)
```

**Comment by seansabour** [comment]
Hi @hagay3 - how difficult would this be to implement? I'd like to take a stab at it if possible.

**Comment by hagay3** [comment]
@seansabour 
I appreciate your interest in skuber
You can look at the following code references for how it's done in GCP.


https://github.com/hagay3/skuber/blob/master/client/src/main/scala/skuber/api/client/package.scala#L102

The difficulty level is based on your understanding of aws eks token lifecycle and the integration with k8s.

In GCP case its done with gcp cli which generates a token every refresh interval.
For eks, I want this feature will be developed with eks java client rather than aws cli, it's not ideal to be dependent on a cli command and it also requires the users of skuber to bundle the aws cli command in their image.

explanation of generating eks token with aws eks java client:
https://stackoverflow.com/questions/59853623/get-authentication-token-from-aws-eks-using-the-aws-java-sdk-v2

Let me know what do you think about that.

**Comment by hagay3** [comment]
@seansabour @brycechesternewman 
I have started to work on this feature
Let me know if you have any suggestions or improvements
https://github.com/hagay3/skuber/pull/163/files

**Comment by hagay3** [comment]
@brycechesternewman 
The feature is ready, I'm struggling with testing it on ECR cluster because I don't have access to a real ECR cluster.
Can you help with that case?
you need to create an IAM role, attach it to service account and then attach it to the pod
full instructions 
https://gist.github.com/hagay3/f81243ca3c86396c2a4064823b37d0c3

the jar for tests can be downloaded from here
https://drive.google.com/file/d/1pCM1mrZPn927uLcJchVFqgjhO-I_3ZEq/view?usp=sharing

here is the code you need to run from CLI inside the pod in order to run the test app, and replace your cluster details in the env variables placeholders.
```
export namespace=default;export serverUrl=https://EKS_URL; export certificate=CERTIFICATE_BASE64_ENCODED;export clusterName=EKS_CLUSTER_NAME; export region=REGION_NAME

java -cp skuber-examples-assembly-2.7.4.jar  skuber.examples.auth.AwsAuthExample
```

**Comment by brycechesternewman** [comment]
Thank you. I should be able to get to test this next week. Though I have some concerns about region of sts I noted in the pull request.

**Comment by hagay3** [comment]
@brycechesternewman 
I have added the fix, also the test jar is updated
also added full instructions for creating the IAM role
https://gist.github.com/hagay3/f81243ca3c86396c2a4064823b37d0c3

**Comment by hagay3** [comment]
@brycechesternewman 
The PR is tested and merged, skuber v2.7.5 was released.
You can find a very detailed guide here [Refresh EKS Token](https://github.com/hagay3/skuber/blob/master/docs/Refresh_EKS_AWS_Token.md)

**Comment by brycechesternewman** [comment]
I finally had chance to look into this. I don't see any problems. Thanks again.
